core/src/main/scala/org/http4s/Entity.scala

```diff
@@ -16,19 +16,64 @@
 
 package org.http4s
 
-import cats.Monoid
+import fs2.{Chunk, Stream}
+import cats._
 import cats.syntax.all._
 
-final case class Entity[+F[_]](body: EntityBody[F], length: Option[Long] = None)
+sealed trait Entity[F[_]] {
+  def body: Stream[F, Byte]
+  def length: Option[Long]
+  def translate[G[_]](fk: F ~> G): Entity[G]
+  def covary[G[x] >: F[x]]: Entity[G] = translate[G](cats.arrow.FunctionK.id[F].widen[G])
+}
 
 object Entity {
+
+  case class Strict[F[_]](chunk: Chunk[Byte]) extends Entity[F] {
+    def body: Stream[F, Byte] = Stream.chunk(chunk)
+    def length: Option[Long] = chunk.size.toLong.some
+    def translate[G[_]](fk: F ~> G): Entity[G] = this.asInstanceOf[Strict[G]]
+  }
+
+  case class TrustMe[F[_]](body: Stream[F, Byte], size: Long) extends Entity[F] {
+    def length: Option[Long] = size.some
```

Comment:
Should we enforce that the size lines up with the stream size by doing `body.take(size)`? I think that would make the behavior more consistent for both client and server implementations.

I suppose the question is whether or not we ever want to permit reading/writing a body which is larger than the declared size. Is that ever valuable?

Reply:
OTOH we don't really gain much by enforcing that either I suppose.

Reply:
Question is, what do we do about it?  Either an underflow or overflow will foul up the next message on a persistent connection.